<a href="https://colab.research.google.com/github/miruthula05-dev/Python/blob/master/Python_Coding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **List Questions**

## 1_find palindrome in a list

In [ ]:
def is_palindrome(lst):
    """
    Function to check if a list is a palindrome.
    :param lst: List[int] -> List of integers
    :return: bool -> True if the list is a palindrome, False otherwise
    """
    new_lst= []
    for i in range(len(lst)-1, -1, -1):
      new_lst.append(lst[i])

    return new_lst == lst

print(is_palindrome([1,2,3,4,3,2,1]))


True


## 2_count transition

In [ ]:
def change_count(lst):
  count = 0

  for i in range(len(lst)-1):
    if lst[i] != lst[i+1]:
      count += 1

  return count

print(change_count([0,1,0,0,1,1,0,0,1,1]))


5


## 3_reverse a list

In [ ]:
def reverse_lst(lst):
    """
    Function to check if a list is a palindrome.
    :param lst: List[int] -> List of integers
    :return: bool -> True if the list is a palindrome, False otherwise
    """
    new_lst= []
    for i in range(len(lst)-1, -1, -1):
      new_lst.append(lst[i])

    return new_lst

print(reverse_lst([1,2,3,4,5]))

[5, 4, 3, 2, 1]


olleh


## 4_increment by one

In [ ]:
def increment_by_one(digits):
    # Start from the last digit (rightmost)
    for i in range(len(digits) - 1, -1, -1):  # Traverse the list from the end to the start
        if digits[i] < 9:  # If the current digit is less than 9
            digits[i] += 1  # Just add 1 to it, no carry needed
            return digits  # Return the updated list, as no further changes are needed
        digits[i] = 0  # If the digit is 9, set it to 0 and carry over 1 to the next digit

    # If we exit the loop, it means we had a carry over past the most significant digit
    return [1] + digits  # Add '1' at the beginning of the list (carry over to a new place)

print(increment_by_one([1,2,3]))

[1, 2, 4]


##  5_Max element in a list

In [ ]:
def max_lst(lst):
  maxi = lst[0]
  for i in lst:
    if i > maxi:
      maxi = i
  return maxi

print(max_lst([1,2,3,4,5]))



5


## 6_Sum of element in a list

In [ ]:
def find_sum(lst):
  sum = 0
  for i in lst:
    sum += i
  return sum

print(find_sum([1,2,3,4,5]))

15


## 7_max consequetive ones

In [ ]:
def find_max_consecutive_ones(nums):
    """
    Function to find the maximum number of consecutive 1's in a binary array.
    :param nums: List[int] -> A binary array where each element is either 0 or 1
    :return: int -> The maximum number of consecutive 1's
    """
    max_count = 0  # To track the maximum number of consecutive 1's
    current_count = 0  # To track the current streak of consecutive 1's

    for num in nums:
        if num == 1:
            current_count += 1  # Increment the current streak if it's 1
            max_count = max(max_count, current_count)  # Update max if current streak is greater
        else:
            current_count = 0  # Reset the streak if 0 is encountered

    return max_count

# Example usage:
print(find_max_consecutive_ones([1, 0, 1, 1, 1]))  # Output should be 3


3


## 8_Remove Duplicate in a List


In [ ]:
li = [1,2,2,3,4,4]

new_li = []

for i in range(0, len(li)):
  if li[i] not in new_li:
    new_li.append(li[i])

new_li



[1, 2, 3, 4]

## 9_Count Number of Odd and Even Elements in a List

In [ ]:
li = [1,2,3,4,5,6,7,8,9]

odd_count = 0
eve_count = 0

for i in range(len(li)):
  if li[i] % 2:
    odd_count += 1
  else:
    eve_count += 1


print("Odd = ", odd_count)
print("even = ", eve_count)

Odd =  5
even =  4


## 10_Maximum difference between two consecutive elements in a list.

In [ ]:
li = [1,2,4,10,90,60]

diff = 0

for i in range(len(li)-1):
  if li[i+1] - li[i] > diff:
    diff = li[i+1] - li[i]

diff

80

## 11_Merge two sorted lists into one sorted list

In [ ]:
li_1 = [1,5,7,9,13]
li_2 = [2,4,6,8,10]

merge_li = []
i = 0
j = 0

while (i < len(li_1) and j < len(li_2)):
  if li_1[i] < li_2[j]:
    merge_li.append(li_1[i])
    i += 1
  else:
    merge_li.append(li_2[j])
    j += 1

while i < len(li_1):
  merge_li.append(li_1[i])
  i += 1

    # Append remaining elements of list2 (if any)
while j < len(li_2):
  merge_li.append(li_2[j])
  j += 1

merge_li


[1, 2, 4, 5, 6, 7, 8, 9, 10, 13]

## 12_Rotate a List

In [ ]:
def rotate_list(lst, k, direction="right"):
    """
    Rotate a list by k positions.

    :param lst: The list to rotate
    :param k: Number of positions to rotate
    :param direction: Direction of rotation ('right' or 'left')
    :return: Rotated list
    """
    r_li = []
    n = len(lst)

    if direction == 'left':

      for i in range(k, n):
        r_li.append(lst[i])

      for i in range(0, k):
        r_li.append(lst[i])

      return r_li

    if direction == 'right':

      for i in range(n-k, n): #[4,1,2,3]
        r_li.append(lst[i])

      for i in range(0, n-k):
        r_li.append(lst[i])

      return r_li

lst = [1,2,3,4]

li = rotate_list(lst, 1, "right")

li




[4, 1, 2, 3]

## 13_Merge 2 List into Dictionary

In [ ]:
dict_1 = {}

li_1 = [1,2,3]
li_2 = ['a','b','c']

min =  len(li_1) if len(li_1) < len(li_2) else len(li_2)

for i in range(min):
  dict_1[li_1[i]] = li_2[i]

dict_1



{1: 'a', 2: 'b', 3: 'c'}

## 14_To check if a list is a subset of another list

In [ ]:
# To check if a list is a subset of another list, you can compare the elements of the first
#  list with the elements of the second list. A subset means all elements of the first list must
#  be present in the second list, but not necessarily in the same order or contiguous.

def is_subset_manual(list1, list2):
    """
    Check if list1 is a subset of list2 without using in-built functions like 'in' or 'set()'.

    :param list1: The first list
    :param list2: The second list
    :return: True if list1 is a subset of list2, False otherwise
    """
    # For each element in list1, check if it is in list2
    for item in list1:
        found = False  # Flag to check if the item is found in list2
        for elem in list2:
            if item == elem:
                found = True
                break  # Break inner loop if item is found
        if not found:
            return False  # Return False if any element is not found in list2
    return True  # Return True if all elements of list1 are found in list2

# Example usage
list1 = [1, 2, 3]
list2 = [1, 2, 3, 4, 5]
print(is_subset_manual(list1, list2))  # Output: True

list3 = [1, 6]
print(is_subset_manual(list3, list2))  # Output: False



## 15_give all the subset of a given list

In [ ]:
def all_subsets(lst):
    result = [[]]  # Start with an empty subset

    for i in lst:
      result += [current + [i] for current in result]
    return result[1:]  # Remove empty subset if not needed

# Example
print(all_subsets([1, 2, 3]))


[[1], [2], [1, 2], [3], [1, 3], [2, 3], [1, 2, 3]]


## 16_Two Sum - Leetcode1

In [ ]:
# Given an array of integers nums and an integer target, return indices of
# the two numbers such that they add up to target.

# You cannot use the same element twice.
# You can return the answer in any order.
# Assume there is exactly one solution.


# nums = [2, 7, 11, 15]
# target = 9
# output = [0, 1]

# Brute Force approach

def two_sum(lst, target):
  for i in range(len(lst)):
    for j in range(i+1, len(lst)):
      if lst[i] + lst[j] == target:
        return i,j

print(two_sum([2, 7, 11, 15], 9))
print(two_sum([3,2,4], 6))

# 🔹 Time Complexity
# O(n²) because we check every pair (i, j).
# Works fine for small lists, but slow for large inputs.

def two_sum_optimized(nums, target):
    num_map = {}  # Dictionary to store (value → index)

    for i, num in enumerate(nums):
        complement = target - num  # Find complement needed

        if complement in num_map:  # If complement exists, return indices
            return [num_map[complement], i]

        num_map[num] = i  # Store the number in hashmap
# Optimized (HashMap)	O(n)


(0, 1)
(1, 2)


## 17_Remove Duplicates from Sorted Array - LeetCode 26

In [ ]:
# Given a sorted array nums, remove the duplicates in-place such that each element appears only once.
# The relative order of the elements must be maintained.

# 👉 You must modify the input array in-place and return the new length of the modified array.

# 🚨 Do not use extra space for another array.

# nums = [0,0,1,1,1,2,2,3,3,4]
# Output:
# 5, nums = [0,1,2,3,4, _, _, _, _, _]

def remove_duplicates(nums):
    if not nums:
        return 0  # Edge case: empty array

    i = 0  # Pointer to store unique elements

    for j in range(1, len(nums)):
        if nums[j] != nums[i]:  # Found a new unique number
            i += 1  # Move pointer forward
            nums[i] = nums[j]  # Overwrite duplicate

    return i + 1  # Length of unique elements


print(remove_duplicates([0,0,1,1,1,2,2,3,3,4]))

# Use two pointers (i for unique elements, j for traversal).
# Modify the array in-place and return the new length.
# Time Complexity: O(n), Space Complexity: O(1).

5


## 18_Majority Element - Leetcode169

In [ ]:
def majority_element(arr):
  dict = {}

  for num in arr:
    if num not in dict:
      dict[num] = 1
    else:
      dict[num] += 1

  n = len(arr)
  for key, value in dict.items():
      if value > n // 2:
          return key  # Return the majority element

  return None

print(majority_element([4,4,2,4,3]))

# Optimized (HashMap)	O(n)	O(n)



4


## 19_Buy sell Leetcode


In [ ]:
def max_profit_brute(prices):
    max_profit = 0  # Initialize max profit

    for i in range(len(prices)):  # Buy day
        for j in range(i + 1, len(prices)):  # Sell day (must be after buy day)
            profit = prices[j] - prices[i]  # Profit if bought at i and sold at j
            max_profit = max(max_profit, profit)  # Update max profit if better

    return max_profit
prices = [7,1,5,3,6,4]
print(max_profit_brute(prices))

# Efficient code

def max_profit(prices):
    if not prices:  # Edge case: empty list
        return 0

    min_price = float('inf')  # Start with a very high value
    max_profit = 0  # Initialize profit to 0

    for price in prices:
        if price < min_price:
            min_price = price  # Update minimum price
        elif price - min_price > max_profit:
            max_profit = price - min_price  # Update max profit if it's better

    return max_profit


## 20_"Contains Duplicate" (LeetCode 217)

In [ ]:
# Given an integer array nums, return True if any value appears at least twice in the array. Otherwise, return False.

def contains_duplicate(arr):
    unique = []  # Stores unique elements

    for num in arr:
        if num in unique:  # If duplicate found, return True
            return True
        unique.append(num)  # Otherwise, add to list

    return False  # If no duplicates, return False

# Test Cases
print(contains_duplicate([1,2,3,1]))  # Output: True
print(contains_duplicate([1,2,3,4]))  # Output: False
print(contains_duplicate([]))  # Output: False
# The worst-case time complexity is O(n²) because num not in unique takes O(n).

def contains_duplicate(arr):
    seen = set()  # Use a set for fast lookups

    for num in arr:
        if num in seen:
            return True  # Found a duplicate
        seen.add(num)  # Store the number

    return False  # No duplicates found

# A set lookup is O(1)(it uses dictionary), so this solution runs in O(n) time instead of O(n²).


## 21_"Intersection of Two Arrays" (LeetCode 349)

In [ ]:
# Given two integer arrays nums1 and nums2, return an array of their intersection.
# Each element in the result must be unique, and you may return the result in any order.

lst1 = [1,2,3]
lst2 = [2]

new = set()

for item in lst1:
  if item in lst2:
    new.add(item)

print(new)


{2}


# Dictionary

## Merge Multiple Dict

In [ ]:

# method 1

# merge_dict = dict1| dict2| dict3

# merge_dict

# method 2

# merge_dict = {**dict1, **dict2, **dict3}
# merge_dict

# method 3

def merge(*dicts):
  merge_dict = {}

  for d in dicts:
    for key, value in d.items():
      merge_dict[key] = value


  print(merge_dict)

dict1 = {1:'a', 2:'b', 3:'c'}
dict2 = {4:'d', 5:'e', 6: 'f'}
dict3 = {7:'g', 8:'h', 9: 'i'}

merge(dict1, dict2, dict3)


{1: 'a', 2: 'b', 3: 'c', 4: 'd', 5: 'e', 6: 'f', 7: 'g', 8: 'h', 9: 'i'}


## Merge Dictionaries with Common Keys

In [ ]:
def merge_dicts_sum_common_keys(dicts):
    """
    Merge multiple dictionaries and sum values for common keys.

    :param dicts: List of dictionaries to merge
    :return: Merged dictionary with summed values for common keys
    """
    merged_dict = {}

    for d in dicts:
        for key, value in d.items():
            merged_dict[key] = merged_dict.get(key, 0) + value  # Sum values for common keys

    return merged_dict

# Example usage
dict1 = {'a': 1, 'b': 2, 'c': 3}
dict2 = {'b': 3, 'c': 4, 'd': 5}
dict3 = {'a': 2, 'c': 1, 'e': 6}

merged_dict = merge_dicts_sum_common_keys([dict1, dict2, dict3])
print(merged_dict)  # Output: {'a': 3, 'b': 5, 'c': 8, 'd': 5, 'e': 6}


def merge_dicts_append_lists_common_keys(dicts):
    """
    Merge multiple dictionaries and append values into lists for common keys.

    :param dicts: List of dictionaries to merge
    :return: Merged dictionary with lists of values for common keys
    """
    merged_dict = {}

    for d in dicts:
        for key, value in d.items():
            if key in merged_dict:
                merged_dict[key].append(value)  # Append value if key already exists
            else:
                merged_dict[key] = [value]  # Start a new list for the key

    return merged_dict

# Example usage
dict1 = {'a': 1, 'b': 2, 'c': 3}
dict2 = {'b': 3, 'c': 4, 'd': 5}
dict3 = {'a': 2, 'c': 1, 'e': 6}

merged_dict = merge_dicts_append_lists_common_keys([dict1, dict2, dict3])
print(merged_dict)  # Output: {'a': [1, 2], 'b': [2, 3], 'c': [3, 4, 1], 'd': [5], 'e': [6]}


# String

## 1_Words Frequency in a Sentence

In [ ]:
sent = "The world is big"

count_dict = {}  # Dictionary to store word frequencies
word = ""        # Temporary variable to build each word

for i in range(len(sent)):
    if sent[i] != ' ':  # If the character is not a space, build the word
        word += sent[i]
    if sent[i] == ' ' or i == len(sent) - 1:  # Space or last character triggers word processing
        if word:  # Only process if word is non-empty
            word = word.lower()  # Convert to lowercase for case-insensitivity
            count_dict[word] = count_dict.get(word, 0) + 1
            word = ""  # Reset word for the next one

print(count_dict)

def word_frequency(sentence):
    """
    Calculate the frequency of words in a sentence.

    :param sentence: Input sentence (string)
    :return: Dictionary with words as keys and their frequencies as values
    """
    # Convert the sentence to lowercase and split into words
    words = sentence.lower().split()

    # Initialize an empty dictionary to store word frequencies
    frequency = {}

    # Count occurrences of each word
    for word in words:
        if word in frequency:
            frequency[word] += 1
        else:
            frequency[word] = 1

    return frequency

# Example usage
sentence = "This is a test. This test is simple, yet effective!"
frequencies = word_frequency(sentence)
print(frequencies)
# Output: {'this': 2, 'is': 2, 'a': 1, 'test': 2, 'simple': 1, 'yet': 1, 'effective': 1}







{'the': 1, 'world': 1, 'is': 1, 'big': 1}


## 2_Palindromic tuple

In [ ]:
def is_palindromic_tuple(tup):
    """
    Check if a given tuple is palindromic.

    :param tup: Tuple to check
    :return: True if palindromic, False otherwise
    """
    return tup == tup[::-1]  # Check if tuple is equal to its reverse

# Example usage
tup1 = (1, 2, 3, 2, 1)
tup2 = (1, 2, 3, 4, 5)

print(is_palindromic_tuple(tup1))  # Output: True
print(is_palindromic_tuple(tup2))  # Output: False


## 3_Reverse a string

In [ ]:
def reverse_string(s):
    """
    Function to return the reversed version of the input string.

    Parameters:
    s (str): The input string to be reversed.

    Returns:
    str: The reversed string.
    """
    new_str = ""
    n = len(s) - 1
    for i in range(n,-1, -1):
      new_str += s[i]

    return new_str


print(reverse_string("priya"))

ayirp


In [ ]:
# Input string
s = "Geeks for Geeks"

# Split the string into words, reverse the list of words, and join them back
reversed_words = ' '.join(s.split()[::-1])

print(reversed_words)


In [ ]:
# using no extra variable and inbuilt functions
def reverse(s):
    s = list(s)  # Convert string to list (since strings are immutable)
    i, j = 0, len(s) - 1

    while i < j:
        s[i], s[j] = s[j], s[i]  # Swap without extra variables
        i += 1
        j -= 1

    return ''.join(s)  # Convert back to string

# Example
s = "hello"
print(reverse(s))  # Output: "olleh"

olleh


## 4_Count Vowels in a string

In [ ]:
def count_vowels(s):
    """
    Function to count the number of vowels in the input string.

    Parameters:
    s (str): The input string to check for vowels.

    Returns:
    int: The count of vowels in the input string.
    """
    vowels = "aeiou"
    count = 0

    for i in s:
      if i in vowels:
        count += 1

    return count

print(count_vowels("priya"))



2


## 5_check two strings are equal at the same position and order

In [ ]:
def are_equal_strings(s, t):
    """
    Function to check if two strings are equal without using built-in functions.

    Parameters:
    s (str): The first string.
    t (str): The second string.

    Returns:
    bool: True if the strings are equal, False otherwise.
    """
    if len(s) != len(t):
      return False

    for i in range(len(s)):
      if s[i] != t[i]:
        return False
    return True

print(are_equal_strings("Hello", "Hello"))

True


## 6_check palindrome of a string

In [ ]:
def is_palindrome(s):
    """
    Function to check if the input string is a palindrome.

    Parameters:
    s (str): The input string to check.

    Returns:
    bool: True if the string is a palindrome, False otherwise.
    """
    s = s.lower().replace(" ", "")

    n = len(s)-1
    new_str = ""

    for i in range(n, -1, -1):
      new_str += s[i]

    return s == new_str

is_palindrome("Madam")






madam


True

In [ ]:
def is_palindrome(s):
    i, j = 0, len(s) - 1

    while i < j:
        if s[i] == '*' or s[j] == '*':
			      pass
        if s[i] != s[j]:  # Compare characters
            return False
        i += 1
        j -= 1

    return True  # If loop completes, it's a palindrome

# Example cases
print(is_palindrome("madam"))   # Output: True
print(is_palindrome("hello"))   # Output: False
print(is_palindrome("racecar")) # Output: True
print(is_palindrome("12321"))   # Output: True


In [ ]:
# New version of palindrome
# Ex : “m*m” .. if u replace * with any word this can be palindrome

# Input is a string which has “*” in it

# For the given word if u replace “*” with any letter can it be formed as palindrome or not

# 1. Function should take string as input 2. Return true if it can be palindrome else false

# Eg : “minio*” this cannot be be palindrome

# Word can have more than one *
def is_palindrome(s):
    i, j = 0, len(s) - 1

    while i < j:
        if s[i] == '*' or s[j] == '*':
			      pass
        elif s[i] != s[j]:  # Compare characters
            return False
        i += 1
        j -= 1

    return True  # If loop completes, it's a palindrome

# Example cases
print(is_palindrome("m*m"))
print(is_palindrome("a*m*"))
print(is_palindrome("minio*"))


True
True
False


## 7_Count words in a string

In [ ]:
def count_words(s):
    """
    Function to count the number of words in the input string.

    Parameters:
    s (str): The input string to check for words.

    Returns:
    int: The count of words in the input string.
    """
    count = 1
    lst =[]
    word = ""
    for i in s:
      if i != ' ':
        word += i
      elif i == ' ':
        if word: # To ensure word is not empty
            lst.append(word)
            word = ""

    if word: # to cature the last word
      lst.append(word)

    return len(lst)

print(count_words("Hello, World"))

Hello,
2


## 8_Remove Duplicates of a string

In [ ]:
def remove_duplicates(s):
    """
    Function to remove duplicate characters from the input string.

    Parameters:
    s (str): The input string from which duplicates need to be removed.

    Returns:
    str: The modified string with duplicates removed.
    """
    new_str = ''
    for i in s:
      if i not in new_str:
        new_str += i

    return new_str

print(remove_duplicates("hello"))


helo


## 9_Count constants in a string

In [ ]:
def count_consonants(s):
    """
    Function to count the number of consonants in the input string.

    Parameters:
    s (str): The input string to check for consonants.

    Returns:
    int: The count of consonants in the input string.
    """
    vowels = "aeiou"
    count = 0
    s  = s.lower().replace(" ","")
    for char in s :
        if ('a' <= char <= 'z'):
          if char not in vowels:
            count += 1
    return count
print(count_consonants("hello"))


3


## 10_Check for anangrams

In [ ]:
def is_anagram(s, t):
    if len(s) != len(t):
        return False  # If lengths are different, they can't be anagrams

    char_count = {}  # Dictionary to count characters

    for char in s:
        char_count[char] = char_count.get(char, 0) + 1  # Count occurrences in s

    for char in t:
        if char not in char_count or char_count[char] == 0:
            return False  # If char doesn't exist or count goes negative, not an anagram
        char_count[char] -= 1  # Reduce count for matching char

    return True  # All counts matched, it's an anagram

# 🔹 Test Cases
print(is_anagram("hello", "ollih"))  # ❌ False (incorrect input)
print(is_anagram("listen", "silent"))  # ✅ True
print(is_anagram("rat", "car"))  # ✅ False
print(is_anagram("aabb", "abab"))  # ✅ True

# Time Complexity: O(n)

False


## 11_Check Subsequence

You are given two strings s and t. Your task is to determine if string t is a subsequence of string s. A subsequence of a string is a new string that is formed from the original string by deleting some (or no) characters without changing the order of the remaining characters.



Input:

Two strings s and t where the length of s is between 1 and 1000, and the length of t is between 1 and 1000.



Output:

Return True if t is a subsequence of s, and False otherwise.



Example:

Input: s = "abcde", t = "ace"
Output: True

Input: s = "abcde", t = "aec"
Output: False


In [ ]:
def is_subsequence(s, t):
    """
    Function to check if t is a subsequence of s.

    Parameters:
    s (str): The original string.
    t (str): The target subsequence string.

    Returns:
    bool: True if t is a subsequence of s, False otherwise.
    """
    i = 0 ; j = 0

    while i < len(s) and j < len(t):
      if s[i] == t[j]:
        j += 1
      i += 1
    return j == len(t)


print(is_subsequence("abcde","ace"))

True


## 12_Check for substring

Problem Description:

You are given two strings, s and t. Your task is to determine if the string t is a substring of the string s. A substring is a contiguous sequence of characters within a string. Do not use any built-in functions for string operations and do not use recursion.



Input:

Two strings s and t, where 1 <= len(s), len(t) <= 1000.



Output:

A boolean value (True or False) indicating whether t is a substring of s.



Example:

Input: s = "hello world", t = "world"
Output: True

Input: s = "hello world", t = "worlds"
Output: False

In [ ]:
def check_substring(s, t):
  return t in s

print(check_substring("Hello, World", "lle"))

False


## 13_length of the longest word

In [ ]:
def longest_word_length(s):
  word = ""
  lst = []
  s = s.lower()
  for i in s:
    if i != ' ':
      word += i
    elif i == ' ':
      lst.append(word)
      word = ''

  if word:
    lst.append(word)

  return max(len(w) for w in lst)

print(longest_word_length("The world is Big"))

##

def longest_word_length(s):
    max_length = 0
    current_length = 0

    for char in s:
        if char != ' ':
            current_length += 1
        else:
            max_length = max(max_length, current_length)
            current_length = 0

    # Check the last word's length
    max_length = max(max_length, current_length)

    return max_length

print(longest_word_length("The world is Big"))  # Output: 5


5


## 14_Check whether a string is palindrome or symmtrical or not

In [ ]:
def is_palindrome(s):
  i = 0
  j= len(s)-1

  while i < j:
    if s[i] != s[j]:
      return False
    i += 1
    j -= 1

  return True

def is_symetrical(s):
  i = 0
  j = len(s)// 2

  while j < len(s):
    if s[i] != s[j]:
      return False
    i += 1
    j += 1

  return True

print(is_palindrome("khokho"))
print(is_symetrical("khokho"))

## 15_Remove i th character from a string

In [ ]:
s = "hello"
k = 2
new_s = ""

for i in range(len(s)):
  if i != k:
    new_s += s[i]

print(new_s)

helo


In [ ]:
s = "hello world"

# Use a list comprehension to create a new string by joining
# characters that are not 'o' from original string 's'
s = "".join([c for c in s if c != "o"])
print(s)


## 16_Convert snake case to pascal case

In [ ]:
s= 'geeksforgeeks_is_best'
res = ''.join(word.capitalize() for word in s.split('_'))
print(res)


GeeksforgeeksIsBest


In [ ]:
s= 'geeksforgeeks_is_best'
new_s = ""
word =""
r = ""
for i in s:
  if i != '_':
    word += i

  if i == '_':
    r = word.capitalize()
    new_s = new_s+ r
    word = ""

if word:
  r = word.capitalize()
  new_s = new_s+ r


print(new_s)

GeeksforgeeksIsBest


In [ ]:
s= 'geeksforgeeks_is_best'
res = ""
capNext = True  # Flag to track if the next character should be capitalized

for char in s:
    if char == '_':
        capNext = True
    elif capNext:
        res += char.upper()  # Capitalize the current character
        capNext = False
    else:
        res += char  # Add the character as it is

print(res)


## 17_Print even length words in a string

In [ ]:
s = 'The world is bigi'

s = " ".join([word for word in s.split() if len(word)% 2 == 0])
print(s)

is bigi


In [ ]:

s = 'The world is bigi'
word = ""
new_s = ""

for i in s:
  if i != " ":
    word += i
  if i == " ":
    if len(word) % 2 == 0:
      new_s += word
    word = ""
if word:
  new_s += " "
  if len(word) % 2 == 0:
      new_s += word


print(new_s)

is bigi


In [ ]:
# Python3 program to print
# even length words in a string

def printWords(s):

	# split the string
	s = s.split(' ')

	# iterate in words of string
	for word in s:

		# if length is even
		if len(word)%2==0:
			print(word)


# Driver Code
s = "i am muskan"
printWords(s)


## 18_Accept the string which contain all vowels

In [ ]:
s = "Geeksforgeeks"
v = 'aeiou'

def check_string_vowels(s, v):

  for i in v:
    if i not in s:
      return False
  return True

check_string_vowels(s, v)

False

##  19_Check a given string is binary or not


In [ ]:
# Input: str = "01010101010"
# Output: Yes

def check_binary(s):
  bin = "01"
  for i in s:
    if i not in bin:
      return False

  return True

check_binary("010p0101010")



False

## 20_Split and join a string

In [ ]:
a = "Hello, how are you?"
new_s = ""

for i in a:
  if i != " ":
    new_s += i
  elif i == " ":
    new_s += "-"


print(new_s)

Hello,-how-are-you?


## 21_program for removing i-th character from a string

In [ ]:
s = "PythonProgramming"

# Index of the character to remove
i = 6

new_s = ""

for k in range(len(s)):
  if k != i:
    new_s += s[k]

print(new_s)


Pythonrogramming


## 22_Find words which are greater than given length k

In [ ]:
# Input : str = "string is fun in python"
#         k = 3
# Output : string python

def find_words(s, k):
  new_s = ""
  word = ""

  for i in s:
    if i != " ":
      word += i
    elif i == " ":
      if len(word) > k:
        new_s += word

      new_s += " "
      word = ""

  if word:
    if len(word) > k:
        new_s += word

  return new_s

find_words("string is fun in python", 3)







'string    python'

## 23_Check if string contains any special ch

In [ ]:
def has_special_char(s):
    for c in s:
        if not (c.isalpha() or c.isdigit() or c == ' '):
            return True
    return False


# Test the function
s = "Hello World"
if has_special_char(s):
    print("The string contains special characters.")
else:
    print("The string does not contain special characters.")

s = "Hello@World"
if has_special_char(s):
    print("The string contains special characters.")
else:
    print("The string does not contain special characters.")
# This code is contributed by Edula Vinay Kumar Reddy


## 24_Maximum frequency character in String

In [ ]:
s = "hello world"

dict_1 = {}
max = 0

for i in s:
  if i not in dict_1:
    dict_1[i] = 1
  else:
    dict_1[i] += 1

for k, v in dict_1.items():

  if v > max:
    max = v

for k, v in dict_1.items():
  if v == max:
    print(k)


l


In [ ]:
from collections import Counter

s = "hello world"
freq_dict = Counter(s)

max_freq = max(freq_dict.values())  # Find max frequency

# Print characters with max frequency
for char, freq in freq_dict.items():
    if freq == max_freq:
        print(char)


In [ ]:
s = "hello world"

# Create a frequency dictionary
freq_dict = {char: s.count(char) for char in set(s)}

max_freq = max(freq_dict.values())  # Find max frequency

# Print characters with max frequency
for char, freq in freq_dict.items():
    if freq == max_freq:
        print(char)


## 25_program to find uncommon words from two Strings

In [ ]:
# Python3 program to find a list of uncommon words

# Function to return all uncommon words


def UncommonWords(A, B):

	# count will contain all the word counts
	count = {}

	# insert words of string A to hash
	for word in A.split():
		count[word] = count.get(word, 0) + 1

	# insert words of string B to hash
	for word in B.split():
		count[word] = count.get(word, 0) + 1

	# return required list of words
	return [word for word in count if count[word] == 1]


# Driver Code
A = "Geeks for Geeks"
B = "Learning from Geeks for Geeks"

# Print required answer
print(UncommonWords(A, B))


In [ ]:
from collections import Counter

def get_uncommon_words(str1, str2):
    def extract_words(s):
        words = []
        word = ""
        for char in s:
            if char.isalnum():  # Check if character is a letter or number
                word += char
            elif word:  # If we hit a space or punctuation and word is non-empty
                words.append(word)
                word = ""
        if word:  # Add the last word if it exists
            words.append(word)
        return words

    # Extract words manually without split()
    words1 = extract_words(str1)
    words2 = extract_words(str2)

    # Count word frequencies
    freq = Counter(words1 + words2)

    # Get uncommon words (appearing only once)
    uncommon_words = [word for word, count in freq.items() if count == 1]

    return uncommon_words

# Example Usage
str1 = "Hello world! This is a test."
str2 = "This test is simple."

print(get_uncommon_words(str1, str2))


# numbers Question

## 1_Add two nos

In [ ]:
# Python3 program to add two numbers

num1 = 15
num2 = 12

# Adding two nos
import operator
su = operator.add(num1,num2)

# printing values
print("Sum of {0} and {1} is {2}" .format(num1,
                                       num2, su))


# Define a lambda function to add two numbers
add_numbers = lambda x, y: x + y

# Take input from the user
num1 = 1
num2 = 2

# Call the lambda function to add the two numbers
result = add_numbers(num1, num2)

# Print the result
print("The sum of", num1, "and", num2, "is", result)


# Define a recursive function to add two numbers
def add_numbers_recursive(x, y):
    if y == 0:
        return x
    else:
        return add_numbers_recursive(x + 1, y - 1)

# Take input from the user
num1 = 1
num2 = 2

# Call the recursive function to add the two numbers
result = add_numbers_recursive(num1, num2)

# Print the result
print("The sum of", num1, "and", num2, "is", result)


## 2_max of two nos

In [ ]:
max = lambda a,b : a if a>b else b

print(max(5,7))

7


## 3_Factorial of a number

In [ ]:
a = 5
sum = 1
while a:
  sum = sum * a
  a = a - 1

print(sum)

#################################################

# Input: An integer number
num = 6

# Initialize the factorial variable to 1
factorial = 1

# Calculate the factorial using a for loop
for i in range(1, num + 1):
    factorial *= i

# Output: The factorial of the number
print(f"The factorial of {num} is {factorial}")

#################################################
# Python 3 program to find
# factorial of given number
def factorial(n):

    # single line to find factorial
    return 1 if (n==1 or n==0) else n * factorial(n - 1)

# Driver Code
num = 5
print("Factorial of",num,"is",factorial(num))



120


## 4_Armstrong number

In [ ]:
def armstrong(num):
  num_original = num
  sum = 0

  while num:
    rem = num % 10
    sum = sum + (rem ** 3)
    num = num // 10
  return num_original == sum

print(armstrong(153))

True


## 5_Sum of even nos

In [ ]:
def sum_of_even_numbers(n):
    """
    Function to return the sum of the first n even natural numbers.

    Parameters:
    n (int): The number of even numbers to sum.

    Returns:
    int: The sum of the first n even natural numbers.
    """
    sum = 0
    for i in range(1, n+1):
      sum += 2*i
    return sum

print(sum_of_even_numbers(5))

30


## 6_Check for even number

In [ ]:
def is_even(n):
    """
    Function to check if a number is even.

    Parameters:
    n (int): The number to check.

    Returns:
    bool: True if n is even, False otherwise.
    """
    if n % 2 == 0:
        return True
    else:
        return False
is_even(5)

False

## 7_check for Prime

In [ ]:
def is_prime(n):
    """
    Function to check if a number is prime.

    Parameters:
    n (int): The number to check.

    Returns:
    bool: True if n is prime, False otherwise.
    """
    if n < 2:
      return False

    for i in range(2, int(n ** 0.5) + 1):
      if n%i == 0:
        return False

    return True

is_prime(3)

True

In [ ]:
def prime(starting_range, ending_range):
    lst = []
    for i in range(starting_range, ending_range):
        flag = 0  # Reset flag for each number
        for j in range(2, i):
            if i % j == 0:
                flag = 1
                break
        if flag == 0:
            lst.append(i)
    return lst


# Driver program
starting_range = 2
ending_range = 7
lst = prime(starting_range, ending_range)
if len(lst) == 0:
    print("There are no prime numbers in this range")
else:
    print("The prime numbers in this range are: ", lst)



The prime numbers in this range are:  [2, 3, 5]


## 8_Fibanacci

In [ ]:
def fib(n):
  a = 0
  b = 1
  c = a+b
  print(b)
  print(c)

  for i in range(2, n):
      a = b
      b = c
      c = a+b
      print(c)

fib(10)


1
1
2
3
5
8
13
21
34
55


## 9_To check a n is a fibanacci

In [ ]:
# python program to check if x is a perfect square
import math

# A utility function that returns true if x is perfect square


def isPerfectSquare(x):
	s = int(math.sqrt(x))
	return s*s == x

# Returns true if n is a Fibonacci Number, else false


def isFibonacci(n):

	# n is Fibonacci if one of 5*n*n + 4 or 5*n*n - 4 or both
	# is a perfect square
	return isPerfectSquare(5*n*n + 4) or isPerfectSquare(5*n*n - 4)


# A utility function to test above functions
for i in range(1, 11):
	if (isFibonacci(i) == True):
		print(i, "is a Fibonacci Number")
	else:
		print(i, "is a not Fibonacci Number ")


## 10_Check perfect square

In [ ]:
def is_perfect_square(n):
    """
    Check if a number is a perfect square.

    Parameters:
    n (int): The number to check.

    Returns:
    bool: True if n is a perfect square, False otherwise.
    """
    if n < 0:
        return False  # Negative numbers cannot be perfect squares

    sqrt = int(n ** 0.5)  # Calculate the square root of n
    return sqrt * sqrt == n  # Check if the square of sqrt equals n

# Test cases
print(is_perfect_square(16))

## 11_Decimal to binary

In [ ]:
def decimal_to_binary(n):
    """
    Convert a decimal number to binary.

    Parameters:
    n (int): The decimal number to convert.

    Returns:
    str: The binary representation of the number.
    """
    if n == 0:
        return "0"

    binary = ""
    while n > 0:
        remainder = n % 2
        binary = str(remainder) + binary
        n = n // 2

    return binary

# Test cases
print(decimal_to_binary(10))  # Output: 1010
print(decimal_to_binary(0))   # Output: 0
print(decimal_to_binary(1))   # Output: 1
print(decimal_to_binary(255)) # Output: 11111111


1010
0
1
11111111


## 12_Binary to Decimal

In [ ]:
def binary_to_decimal(binary_str):
    """
    Convert a binary number (string) to decimal without using built-in int() function.

    Parameters:
    binary_str (str): The binary number as a string.

    Returns:
    int: The decimal equivalent of the binary number.
    """
    decimal = 0
    length = len(binary_str)

    # Iterate over the binary string and convert to decimal
    for i in range(length):
        # Get the bit at the current position (0 or 1)
        bit = int(binary_str[i])

        # Add the value of the bit multiplied by the corresponding power of 2
        decimal += bit * (2 ** (length - i - 1))

    return decimal

# Test cases
print(binary_to_decimal("1010"))   # Output: 10
print(binary_to_decimal("11111111"))  # Output: 255
print(binary_to_decimal("0"))     # Output: 0
print(binary_to_decimal("1"))     # Output: 1



10
255
0
1


## 13_GCD

In [ ]:
def gcd(a, b):
    """
    Function to find the GCD of two numbers using the Euclidean algorithm.

    Parameters:
    a, b (int): The two numbers whose GCD is to be calculated.

    Returns:
    int: The GCD of the two numbers.
    """
    while b != 0:
      a, b = b, a%b

    return a

print(gcd(101, 10))  # Output: 1
print(gcd(36, 60))  # Output: 12


1
12


## 14_Print ascii value of a string


In [ ]:
str = "priya"

for ch in str:
  print(ord(ch))

112
114
105
121
97


## 15_sum of squares of first n natural numbers

In [ ]:
sum = 0
n = 5

for i in range(n):
  sum = sum + i**2

print(sum)

30


# Bits Related Question

Signed integer  $-2^n-1  to  {2^(n-1)} – 1$

The first bit in a signed representation is the sign of the number, 0 for non-negative numbers and 1 for negative numbers and the remaining n−1 bits contain the magnitude of the number.

Two’s complement is used, which means that the opposite number of a number is calculated by first inverting all the bits in the number, and then increasing the number by one.
The bit representation of the int number −43 is 11111111111111111111111111010101

signed integer

An unsigned variable of n bits can contain any integer between 0 and $2^n −1$.

A signed number −x equals an unsigned number $2^n − x$.
For example, the following pseudo-code snippet shows that the signed number
x = −43 equals the unsigned number y = $2^(32) −43$

1. The left-shift and right-shift operators should not be used for negative numbers.
2. The bitwise OR of two numbers is simply the sum of those two numbers if there is no carry involved; otherwise, you add their bitwise AND.
Let’s say, we have a=5(101) and b=2(010), since there is no carry involved, their sum is just a|b. Now, if we change ‘b’ to 6 which is 110 in binary, their sum would change to a|b + a&b since there is a carry involved.
3. The bitwise XOR operator is the most useful operator from a technical interview perspective.
5. The left-shift and right-shift operators are equivalent to multiplication and division by 2 respectively.
6. The & operator can be used to quickly check if a number is odd or even.
7. The ~ operator should be used carefully.

The result of the ~ operator on a small number can be a big number if the result is stored in an unsigned variable. The result may be a negative number if the result is stored in a signed variable (assuming that the negative numbers are stored in 2’s complement form where the leftmost bit is the sign bit).

1. Count the number of 1s in the binary representation of a number.
2. Check if a number is a power of two.
3. Find the only non-repeating element in an array where every other element repeats twice.
4. Find two non-repeating elements in an array where every other element repeats twice.
5. Check if a number has alternating bits (e.g., 101010).
6. Reverse the bits of a given 32-bit unsigned integer.
7. Count the number of bits to flip to convert one number to another.
8. Find the position of the rightmost set bit in a number.
9. Determine if a number is odd or even using bit manipulation.
10. Swap two numbers without using a temporary variable.
11. Find the missing number in an array containing numbers from 1 to N, with one missing.
12. Calculate the XOR of all elements in a given range [L, R].
13. Generate all possible subsets of a given set using bits.
17. Check if two numbers have opposite signs.
18. Implement addition of two numbers without using the + operator.
19. Find the parity (odd or even) of a number of set bits in a binary representation.
20. Clear the least significant set bit of a number.

# Reg Ex Related Question

## 1_Check if String Contain Only Defined Characters using Regex

In [ ]:
import re
text = '657'
pattern = re.compile('^[87654]+$')

if re.match(pattern, text):
  print("Valid")
else:
  print("invalid")

Valid


## 2_Python program to Count Uppercase, Lowercase, special character and numeric values using Regex

In [ ]:
import re


string = "ThisIsGeeksforGeeks !, 123"

# Creating separate lists using
# the re.findall() method.
uppercase_characters = re.findall(r"[A-Z]", string)
lowercase_characters = re.findall(r"[a-z]", string)
numerical_characters = re.findall(r"[0-9]", string)
special_characters = re.findall(r"[, .!?]", string)

print("The no. of uppercase characters is", len(uppercase_characters))
print("The no. of lowercase characters is", len(lowercase_characters))
print("The no. of numerical characters is", len(numerical_characters))
print("The no. of special characters is", len(special_characters))


## 3_The most occurring number in a string using Regex in python

In [ ]:
word = 'geek55of55gee4ksabc3dr2x'

arr = re.findall(r'[0-9]+', word)
maxi = 0

dict_1 = {}

for i in arr:
  if i in dict_1:
    dict_1[i] += 1
  else:
    dict_1[i] = 1

for k, v in dict_1.items():

  if v > maxi:
    maxi = v

for k, v in dict_1.items():
  if v == maxi:
    print(k)

55


## 4_Find max no in a string

In [ ]:
import re

s = "The price is 120 dollars, and the discount is 50, saving 70 more."
#max_value
list_1 =  re.findall(r'\d+', s)
print(max)
m = max([int(num) for num in list_1])
print(m)


2


TypeError: 'int' object is not callable

## 5_Python to put spaces between words starting with capital letters

In [ ]:
import re

def putSpace(input):

	# regex [A-Z][a-z]* means any string starting
	# with capital character followed by many
	# lowercase letters
	words = re.findall('[A-Z][a-z]*', input)

	# Change first letter of each word into lower
	# case
	for i in range(0,len(words)):
	  words[i]=words[i][0].lower()+words[i][1:]
	print(' '.join(words))


# Driver program
if __name__ == "__main__":
	input = 'BruceWayneIsBatman'
	putSpace(input)


bruce wayne is batman


## 6_Check whether a string starts and ends with the same character or not

In [ ]:
# Python program to check if a string starts
# and ends with the same character

# import re module as it provides
# support for regular expressions
import re

# the regular expression
regex = r'^[a-z]$|^([a-z]).*\1$'

# function for checking the string
def check(string):

	# pass the regular expression
	# and the string in the search() method
	if(re.search(regex, string)):
		print("Valid")
	else:
		print("Invalid")

if __name__ == '__main__' :

	sample1 = "abba"
	sample2 = "a"
	sample3 = "abcd"

	check(sample1)
	check(sample2)
	check(sample3)


## 7_ Find one upper case followed by lower case

In [ ]:
import re

def findSequence(text):
      # regex pattern
    pattern = '[A-Z][a-z]+'

    # Use re.search to find any match
    if re.search(pattern, text):
        return 'Yes'
    else:
        return 'No'

# Driver program
if __name__ == "__main__":
    word = 'geeAkAA55of55gee4ksabc3Ar2x'
    print(findSequence(word))
    word = 'Geeks'
    print(findSequence(word))
    word = 'geeksforgeeks'
    print(findSequence(word))


## 8_Remove duplicate words in a sentence

In [ ]:
# Python program to remove duplicate words
# using Regular Expression or ReGex.
import re


# Function to validate the sentence
# and remove the duplicate words
def removeDuplicateWords(input):

    # Regex to matching repeated words
    regex = r'\b(\w+)(?:\W+\1\b)+'

    return re.sub(regex, r'\1', input, flags=re.IGNORECASE)


# Driver Code

# Test Case: 1
str1 = "Good bye bye world world"
print(removeDuplicateWords(str1))

# Test Case: 2
str2 = "Ram went went to to his home"
print(removeDuplicateWords(str2))

# Test Case: 3
str3 = "Hello hello world world"
print(removeDuplicateWords(str3))

# This code is contributed by yuvraj_chandra


## 9_Remove all characters except letters and numbers

In [ ]:
import re

# Input string
s1 = "Hello, World! 123 @Python$"

# Remove all characters except letters and numbers using regex
s2 = re.sub(r'[^a-zA-Z0-9]', '', s1)

# Print the cleaned string
print(s2)


# File Handling Question

# Linked List Related Question

In [ ]:
class Node:
  def __init__(self, data=None, next=None):
    self.data = data
    self.next = next

class Linkedlist:
  def __init__(self):
    self.head = None

  def insert_at_begin(self, data):
    node = Node(data, self.head) # the current head become next pointer since we add at the begining
    self.head = node

  def insert_at_end(self, data):
    if self.head is None:
      self.head = Node(data, None)
    else:
      itr = self.head

      while itr.next:
        itr = itr.next

      itr.next = Node(data, None)

  def insert_values(self, data_list):
    self.head = None
    for data in data_list:
      self.insert_at_end(data)

  def get_length(self):
    count =0
    itr = self.head
    while itr:
      itr = itr.next
      count += 1
    return count

  def remove_at(self, index):
    if index < 0 or index >= self.get_length():
      raise Exception("Invalid Index")
    if index == 0:
      self.head = self.head.next
      return

    count = 0
    itr = self.head
    while itr:
      if count == index-1:
        itr.next = itr.next.next
        break
      itr = itr.next
      count += 1

  def insert_at(self, index, data):
    if index < 0 or index >= self.get_length():
      raise Exception("Invalid Index")
    if index == 0:
      self.insert_at_begin(data)
      return

    count = 0
    itr = self.head
    while itr:
      if count == index-1:
        node = Node(data, itr.next)
        itr.next = node
        break

      itr = itr.next
      count += 1



  def print(self):
    itr = self.head
    llstr = ""

    while itr:
      llstr += str(itr.data) + "-->"
      itr = itr.next

    print(llstr)

ll = Linkedlist()
# ll.insert_at_begin(5)
# ll.insert_at_begin(6)
# ll.insert_at_end(1)
# ll.insert_at_end(2)
# print(ll.get_length())
ll.insert_values(['G','O','R'])
ll.insert_at(0,'S')
ll.print()

S-->G-->O-->R-->


## 2_Reverse a linked list

## 3_Detect a cycle in the linked list

## 4_Find the middle of a linked list

## 5_Merge two sorted linked_list

## 6_Remove duplicates in a linked lsit

## 7_Find the Intersection Point of Two Linked Lists

## 8_Check if a Linked List is a Palindrome

# Stack Related Question

## 1_Implement Stack without using inbuilt function

In [ ]:
class stack:
    def __init__(self):
        self.data = []

    def pop(self):
        if not self.data:
            raise IndexError("Pop from empty stack")  # Raise error if empty
        return self.data.pop()

    def peek(self):
        if not self.data:
            return None  # Return None if empty instead of causing an error
        return self.data[-1]

    def push(self, value):
        self.data.append(value)

    def is_empty(self):
        return len(self.data) == 0

    def size(self):
        return len(self.data)


st = stack()

st.push(1)
st.push(2)
st.push(3)
print(st.peek())       # Expected output: 3
print(st.is_empty())   # Expected output: False
print(st.size())       # Expected output: 3
print(st.pop())        # Expected output: 3
print(st.pop())        # Expected output: 2
print(st.pop())        # Expected output: 1
print(st.pop())        # Raises IndexError: Pop from empty stack


None
False
3


## 2_Balancing Equation

In [ ]:
# Given a string s containing just the characters '(', ')', '{', '}', '[', and ']',
# determine if the input string is valid.

# A string is valid if:

# Open brackets must be closed by the same type of brackets.
# Open brackets must be closed in the correct order.
# Every close bracket has a corresponding open bracket of the same type.

In [ ]:
def is_valid_brute(s):
    while True:
        new_s = s.replace("()", "").replace("{}", "").replace("[]", "")
        if new_s == s:  # If no change, stop looping
            break
        s = new_s  # Update s to the new string
    return s == ""  # If empty, it's valid

# Test cases
# print(is_valid_brute("[("))       # Output: False
# print(is_valid_brute("{[()]}"))   # Output: True
print(is_valid_brute("()[]{}"))   # Output: True
# print(is_valid_brute("([)]"))     # Output: False


1
1
True


In [ ]:
def is_balanced(expr):
    stack = []
    mapping = {')': '(', '}': '{', ']': '['}

    for char in expr:
        if char in mapping:
            top_element = stack.pop() if stack else '#'
            if mapping[char] != top_element:
                return False
        else:
            stack.append(char)

    return not stack

# Example usage
print(is_balanced("{[()]}"))  # Output: True
print(is_balanced("{[(])}"))  # Output: False


## 3_ using deque

In [ ]:
from collections import deque

class Stack:
    def __init__(self):
        self.stack = deque()

    def push(self, item):
        self.stack.append(item)

    def pop(self):
        return self.stack.pop() if self.stack else "Stack is empty"

    def peek(self):
        return self.stack[-1] if self.stack else "Stack is empty"

    def is_empty(self):
        return not self.stack

# Example usage
s = Stack()
s.push(5)
s.push(15)
print(s.pop())  # Output: 15


Why is deque Better than list?
Avoids Resizing Overhead:

list is implemented as a dynamic array, so when it grows beyond capacity, Python needs to allocate new memory and copy elements, which can be expensive.
deque is implemented as a doubly linked list, so it doesn’t require resizing.
Efficient Insertions and Deletions from Both Ends:

list.pop(0) (removing from the front) is O(n) since all elements must shift left.
deque.popleft() is O(1) since it directly removes from the front.
Better for Large Data Operations:

deque avoids memory reallocation issues seen in large list operations.

When to Use What?
✅ Use deque if you need fast appends and pops from both ends.
✅ Use list if you only push and pop from the end and need random access (O(1) indexing).

For a stack, both list and deque provide O(1) push/pop at the end.
Deque is better if you need more flexible insertions/removals.
Deque avoids list resizing overhead, making it slightly more efficient in certain cases.


## 4_Next Greater Element

In [ ]:
def next_greater_brute(arr):
    n = len(arr)
    result = []

    for i in range(n):
        next_greater = -1
        for j in range(i + 1, n):  # Look for greater elements on the right
            if arr[j] > arr[i]:
                next_greater = arr[j]
                break
        result.append(next_greater)

    return result

# Test
print(next_greater_brute([4, 5, 2, 10, 8]))
# Output: [5, 10, 10, -1, -1]


In [ ]:
def next_greater_element(arr):
    stack = []
    result = [-1] * len(arr)

    for i in range(len(arr) - 1, -1, -1):
        while stack and stack[-1] <= arr[i]:
            stack.pop()
        result[i] = stack[-1] if stack else -1
        stack.append(arr[i])

    return result

# Example usage
print(next_greater_element([4, 5, 2, 25]))  # Output: [5, 25, 25, -1]


## 5_Reverse a string using stack

In [ ]:
def reverse_string(s):
    stack = list(s)
    reversed_str = ""

    while stack:
        reversed_str += stack.pop()

    return reversed_str

# Example usage
print(reverse_string("hello"))  # Output: "olleh"


# other misc Questions

Two Sum (Leetcode 1) – Find two numbers in a list that add up to a target.

Remove Duplicates from Sorted Array (Leetcode 26) – Remove duplicates from a sorted list in-place.

Remove Element (Leetcode 27) – Remove a specific element from a list in-place.

Maximum Subarray (Leetcode 53) – Find the contiguous subarray with the largest sum (Kadane’s Algorithm).

Merge Sorted Array (Leetcode 88) – Merge two sorted lists in-place.

Best Time to Buy and Sell Stock (Leetcode 121) – Find the maximum profit by buying and selling a stock once.

Contains Duplicate (Leetcode 217) – Check if a list contains duplicate elements.

Intersection of Two Arrays (Leetcode 349) – Find the common elements between two lists.

Move Zeroes (Leetcode 283) – Move all zeroes in a list to the end while maintaining order.

Plus One (Leetcode 66) – Increment a number represented as a list of digits.

Valid Palindrome (Leetcode 125) – Check if a string is a palindrome (ignoring cases and non-alphanumeric characters).

Reverse String (Leetcode 344) – Reverse a string in-place.

Reverse Words in a String III (Leetcode 557) – Reverse each word in a sentence while keeping their order.

Implement strStr() (Leetcode 28) – Find the first occurrence of a substring inside another string (similar to Python’s .find()).

Longest Common Prefix (Leetcode 14) – Find the longest common prefix among a list of strings.

Valid Anagram (Leetcode 242) – Check if two strings are anagrams (contain the same characters with the same frequency).

First Unique Character in a String (Leetcode 387) – Find the first non-repeating character and return its index.

To Lower Case (Leetcode 709) – Convert all uppercase letters in a string to lowercase.

Defanging an IP Address (Leetcode 1108) – Replace dots (.) in an IP address with [.].

Shuffle String (Leetcode 1528) – Reorder characters in a string based on given indices.

If you're preparing for **LeetCode interviews in Python**, here are some of the **most commonly asked LeetCode questions**, categorized by topic:  

---

### **1️⃣ Arrays & Hashing**  
1. **[Two Sum](https://leetcode.com/problems/two-sum/) (Leetcode 1)** – Find two numbers that sum to a target. *(HashMap solution in O(n))*  
2. **[Contains Duplicate](https://leetcode.com/problems/contains-duplicate/) (Leetcode 217)** – Check if an array contains duplicates.  
3. **[Valid Anagram](https://leetcode.com/problems/valid-anagram/) (Leetcode 242)** – Check if two strings are anagrams.  
4. **[Group Anagrams](https://leetcode.com/problems/group-anagrams/) (Leetcode 49)** – Group words that are anagrams. *(Use defaultdict & sorting!)*  
5. **[Top K Frequent Elements](https://leetcode.com/problems/top-k-frequent-elements/) (Leetcode 347)** – Find the `k` most frequent elements. *(Heap or Bucket Sort solution!)*  

---

### **2️⃣ Two Pointers**  
6. **[Valid Palindrome](https://leetcode.com/problems/valid-palindrome/) (Leetcode 125)** – Check if a string is a palindrome, ignoring non-alphanumeric characters.  
7. **[Two Sum II - Input Array Is Sorted](https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/) (Leetcode 167)** – Find two numbers in a sorted array that sum to a target. *(Two-pointer technique!)*  
8. **[3Sum](https://leetcode.com/problems/3sum/) (Leetcode 15)** – Find all unique triplets that sum to zero.  
9. **[Container With Most Water](https://leetcode.com/problems/container-with-most-water/) (Leetcode 11)** – Maximize water container size using two pointers.  

---

### **3️⃣ Sliding Window**  
10. **[Best Time to Buy and Sell Stock](https://leetcode.com/problems/best-time-to-buy-and-sell-stock/) (Leetcode 121)** – Find the max profit by buying and selling once. *(Single-pass O(n) solution!)*  
11. **[Longest Substring Without Repeating Characters](https://leetcode.com/problems/longest-substring-without-repeating-characters/) (Leetcode 3)** – Find the length of the longest unique substring. *(Sliding window with HashSet!)*  
12. **[Permutation in String](https://leetcode.com/problems/permutation-in-string/) (Leetcode 567)** – Check if a string’s permutation exists in another string. *(Sliding window with frequency count!)*  
13. **[Minimum Window Substring](https://leetcode.com/problems/minimum-window-substring/) (Leetcode 76) 🔥** – Hard but **very common** question! *(Sliding window with two maps!)*  

---

### **4️⃣ Stack / Monotonic Stack**  
14. **[Valid Parentheses](https://leetcode.com/problems/valid-parentheses/) (Leetcode 20)** – Check if a string contains valid parentheses. *(Stack-based solution!)*  
15. **[Min Stack](https://leetcode.com/problems/min-stack/) (Leetcode 155)** – Implement a stack that supports `min()` in O(1).  
16. **[Daily Temperatures](https://leetcode.com/problems/daily-temperatures/) (Leetcode 739)** – Find the next warmer temperature. *(Monotonic Stack!)*  

---

### **5️⃣ Binary Search**  
17. **[Binary Search](https://leetcode.com/problems/binary-search/) (Leetcode 704)** – Implement binary search.  
18. **[Search in Rotated Sorted Array](https://leetcode.com/problems/search-in-rotated-sorted-array/) (Leetcode 33)** – Find a target in a rotated sorted array. *(Modified Binary Search!)*  
19. **[Find Minimum in Rotated Sorted Array](https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/) (Leetcode 153)** – Find the minimum element in a rotated sorted array.  

---

### **6️⃣ Linked List**  
20. **[Reverse a Linked List](https://leetcode.com/problems/reverse-linked-list/) (Leetcode 206)** – Reverse a linked list iteratively and recursively.  
21. **[Merge Two Sorted Lists](https://leetcode.com/problems/merge-two-sorted-lists/) (Leetcode 21)** – Merge two sorted linked lists.  
22. **[Cycle Detection in Linked List](https://leetcode.com/problems/linked-list-cycle/) (Leetcode 141)** – Detect if a linked list has a cycle. *(Use Floyd’s Cycle Detection - Fast & Slow pointers!)*  

---

### **7️⃣ Trees & Graphs**  
23. **[Invert Binary Tree](https://leetcode.com/problems/invert-binary-tree/) (Leetcode 226)** – Flip a binary tree upside down.  
24. **[Maximum Depth of Binary Tree](https://leetcode.com/problems/maximum-depth-of-binary-tree/) (Leetcode 104)** – Find the maximum depth of a binary tree. *(DFS/BFS approach!)*  
25. **[Lowest Common Ancestor of a BST](https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-search-tree/) (Leetcode 235)** – Find the lowest common ancestor in a BST.  

---

### **8️⃣ Dynamic Programming (Easy & Medium)**  
26. **[Climbing Stairs](https://leetcode.com/problems/climbing-stairs/) (Leetcode 70)** – Solve using Fibonacci-like DP.  
27. **[House Robber](https://leetcode.com/problems/house-robber/) (Leetcode 198)** – Solve using DP to maximize sum without robbing adjacent houses.  

---

### **🔥 Must-Know Hard Problems (If You Have Time!)**  
✅ **[Merge Intervals](https://leetcode.com/problems/merge-intervals/) (Leetcode 56)** – Merge overlapping intervals.  
✅ **[Word Break](https://leetcode.com/problems/word-break/) (Leetcode 139)** – Check if a word can be segmented using a dictionary.  
✅ **[Trapping Rain Water](https://leetcode.com/problems/trapping-rain-water/) (Leetcode 42)** – Classic two-pointer/stack problem.  



Here are the **best** LeetCode coding questions for **Lists (Arrays) and Strings**, categorized by difficulty and concept.  

---

## **🔥 Easy Level Problems (Beginner-Friendly)**
### **Arrays (Lists)**
1. **[Two Sum](https://leetcode.com/problems/two-sum/) (Leetcode 1)** – Find two numbers that sum to a target (Use HashMap for O(n) solution).  
2. **[Remove Duplicates from Sorted Array](https://leetcode.com/problems/remove-duplicates-from-sorted-array/) (Leetcode 26)** – Remove duplicates in-place.  
3. **[Remove Element](https://leetcode.com/problems/remove-element/) (Leetcode 27)** – Remove all occurrences of a specific element.  
4. **[Merge Sorted Array](https://leetcode.com/problems/merge-sorted-array/) (Leetcode 88)** – Merge two sorted arrays in-place.  
5. **[Best Time to Buy and Sell Stock](https://leetcode.com/problems/best-time-to-buy-and-sell-stock/) (Leetcode 121)** – Find the max profit by buying and selling a stock once.  
6. **[Contains Duplicate](https://leetcode.com/problems/contains-duplicate/) (Leetcode 217)** – Check if an array contains duplicates.  
7. **[Move Zeroes](https://leetcode.com/problems/move-zeroes/) (Leetcode 283)** – Move all zeroes to the end while maintaining order.  
8. **[Plus One](https://leetcode.com/problems/plus-one/) (Leetcode 66)** – Increment a number represented as a list of digits.  
9. **[Intersection of Two Arrays](https://leetcode.com/problems/intersection-of-two-arrays/) (Leetcode 349)** – Find common elements in two arrays.  
10. **[Majority Element](https://leetcode.com/problems/majority-element/) (Leetcode 169)** – Find the element that appears more than `n/2` times.  

### **Strings**
11. **[Valid Palindrome](https://leetcode.com/problems/valid-palindrome/) (Leetcode 125)** – Check if a string is a palindrome.  
12. **[Reverse String](https://leetcode.com/problems/reverse-string/) (Leetcode 344)** – Reverse a string in-place.  
13. **[Reverse Words in a String III](https://leetcode.com/problems/reverse-words-in-a-string-iii/) (Leetcode 557)** – Reverse each word in a sentence while keeping their order.  
14. **[Implement strStr()](https://leetcode.com/problems/implement-strstr/) (Leetcode 28)** – Find the first occurrence of a substring inside another string.  
15. **[Longest Common Prefix](https://leetcode.com/problems/longest-common-prefix/) (Leetcode 14)** – Find the longest common prefix in an array of strings.  
16. **[Valid Anagram](https://leetcode.com/problems/valid-anagram/) (Leetcode 242)** – Check if two strings are anagrams.  
17. **[First Unique Character in a String](https://leetcode.com/problems/first-unique-character-in-a-string/) (Leetcode 387)** – Find the first non-repeating character.  
18. **[To Lower Case](https://leetcode.com/problems/to-lower-case/) (Leetcode 709)** – Convert all uppercase letters to lowercase.  
19. **[Defanging an IP Address](https://leetcode.com/problems/defanging-an-ip-address/) (Leetcode 1108)** – Replace `.` in an IP address with `[.]`.  
20. **[Shuffle String](https://leetcode.com/problems/shuffle-string/) (Leetcode 1528)** – Rearrange characters in a string based on given indices.  

---

## **🚀 Medium Level Problems (Interview-Level)**
### **Arrays (Lists)**
21. **[3Sum](https://leetcode.com/problems/3sum/) (Leetcode 15)** – Find all unique triplets that sum to zero.  
22. **[Container With Most Water](https://leetcode.com/problems/container-with-most-water/) (Leetcode 11)** – Maximize water container using two pointers.  
23. **[Product of Array Except Self](https://leetcode.com/problems/product-of-array-except-self/) (Leetcode 238)** – Return an array where `result[i]` is the product of all elements except `nums[i]`.  
24. **[Maximum Subarray](https://leetcode.com/problems/maximum-subarray/) (Leetcode 53)** – Find the contiguous subarray with the largest sum (Kadane’s Algorithm).  
25. **[Sort Colors](https://leetcode.com/problems/sort-colors/) (Leetcode 75)** – Sort an array containing 0s, 1s, and 2s (Dutch National Flag Algorithm).  
26. **[Find the Duplicate Number](https://leetcode.com/problems/find-the-duplicate-number/) (Leetcode 287)** – Find the duplicate number in an array.  
27. **[Spiral Matrix](https://leetcode.com/problems/spiral-matrix/) (Leetcode 54)** – Return elements of a matrix in spiral order.  
28. **[Subarray Sum Equals K](https://leetcode.com/problems/subarray-sum-equals-k/) (Leetcode 560)** – Find the number of subarrays whose sum equals `k`.  

### **Strings**
29. **[Longest Substring Without Repeating Characters](https://leetcode.com/problems/longest-substring-without-repeating-characters/) (Leetcode 3)** – Find the length of the longest unique substring.  
30. **[Group Anagrams](https://leetcode.com/problems/group-anagrams/) (Leetcode 49)** – Group words that are anagrams.  
31. **[Multiply Strings](https://leetcode.com/problems/multiply-strings/) (Leetcode 43)** – Multiply two numbers represented as strings.  
32. **[Integer to Roman](https://leetcode.com/problems/integer-to-roman/) (Leetcode 12)** – Convert an integer to a Roman numeral.  
33. **[Roman to Integer](https://leetcode.com/problems/roman-to-integer/) (Leetcode 13)** – Convert a Roman numeral to an integer.  
34. **[Valid Parentheses](https://leetcode.com/problems/valid-parentheses/) (Leetcode 20)** – Check if a string contains valid parentheses.  
35. **[Decode String](https://leetcode.com/problems/decode-string/) (Leetcode 394)** – Decode an encoded string following the pattern `"3[a2[c]]"` → `"accaccacc"`.  
36. **[Minimum Window Substring](https://leetcode.com/problems/minimum-window-substring/) (Leetcode 76) 🔥** – Find the smallest substring containing all characters of another string.  
37. **[Permutation in String](https://leetcode.com/problems/permutation-in-string/) (Leetcode 567)** – Check if a permutation of `s1` exists in `s2`.  

---

## **💀 Hard Level Problems (Advanced Interviews)**
### **Arrays (Lists)**
38. **[Trapping Rain Water](https://leetcode.com/problems/trapping-rain-water/) (Leetcode 42)** – Calculate trapped water between bars (Two-pointer technique).  
39. **[Median of Two Sorted Arrays](https://leetcode.com/problems/median-of-two-sorted-arrays/) (Leetcode 4)** – Find the median of two sorted arrays (Binary search-based approach).  
40. **[First Missing Positive](https://leetcode.com/problems/first-missing-positive/) (Leetcode 41)** – Find the smallest missing positive integer.  

### **Strings**
41. **[Palindrome Partitioning](https://leetcode.com/problems/palindrome-partitioning/) (Leetcode 131)** – Partition a string into palindromic substrings.  
42. **[Edit Distance](https://leetcode.com/problems/edit-distance/) (Leetcode 72)** – Find the minimum number of operations to convert one string into another.  
43. **[Word Ladder](https://leetcode.com/problems/word-ladder/) (Leetcode 127)** – Find the shortest transformation sequence between two words.  

---


# Sorting

## Bubble Sort

In [ ]:
def bubble_sort(arr):
    n = len(arr)
    # Traverse through all elements in the list
    for i in range(n):
        # Last i elements are already sorted, so we don't need to check them
        for j in range(0, n-i-1):
            # Swap if the element found is greater than the next element
            if arr[j] > arr[j+1]:
                arr[j], arr[j+1] = arr[j+1], arr[j]

# Example usage
arr = [64, 34, 25, 12, 22, 11, 90]
bubble_sort(arr)
print("Sorted array:", arr)


i= 0
J=  0
J=  1
J=  2
J=  3
J=  4
J=  5
i= 1
J=  0
J=  1
J=  2
J=  3
J=  4
i= 2
J=  0
J=  1
J=  2
J=  3
i= 3
J=  0
J=  1
J=  2
i= 4
J=  0
J=  1
i= 5
J=  0
i= 6
Sorted array: [11, 12, 22, 25, 34, 64, 90]


## Selection sort

In [ ]:
def selection_sort(arr):
    n = len(arr)

    # Move through the entire array
    for i in range(n):
        # Find the minimum element in the unsorted part of the array
        min_index = i
        for j in range(i+1, n):
            if arr[j] < arr[min_index]:
                min_index = j

        # Swap the found minimum element with the first element of the unsorted part
        arr[i], arr[min_index] = arr[min_index], arr[i]

# Example usage
arr = [64, 34, 25, 12, 22, 11, 90]
selection_sort(arr)
print("Sorted array:", arr)

Sorted array: [11, 12, 22, 25, 34, 64, 90]


In [ ]:
def is_palindrome(s):
  i = 0
  j= len(s)-1

  while i < j:
    if s[i] != s[j]:
      return False
    i += 1
    j -= 1

  return True

def is_symetrical(s):
  i = 0
  j = len(s)// 2

  while j < len(s):
    if s[i] != s[j]:
      return False
    i += 1
    j += 1

  return True

print(is_palindrome("khokho"))
print(is_symetrical("khokho"))





False
True
